### PR TITLE
New matrix and vector initialization methods

### PR DESCRIPTION
**Pull request by s-trinh**
Related #526

- Proof of concept.
- Add reshape operation
- Warning: there are already reshape operations in `vpColVector` and `vpRowVector`:

https://github.com/lagadic/visp/blob/3d93fa772e38eff27297f06ad0b98532ba3c612d/modules/core/include/visp3/core/vpColVector.h#L237

https://github.com/lagadic/visp/blob/3d93fa772e38eff27297f06ad0b98532ba3c612d/modules/core/include/visp3/core/vpRowVector.h#L207

- That means that reshape in `vpColVector` will hide the one from `vpArray2D` with some possible [issues](https://isocpp.org/wiki/faq/strange-inheritance#hiding-rule).

---

Example code:

```
#ifdef VISP_HAVE_CXX11
  {
    vpArray2D<float> a{ 1.f, 2.f, 3.f };
    std::cout << "a:\n" << a << std::endl;
    a = { -1, -2, -3, 4, 5.5, 6.0f };
    std::cout << "a:\n" << a << std::endl;
    a.reshape(2, 3);
    std::cout << "a.reshape(2,3):\n" << a << std::endl;
    a.reshape(3, 2);
    std::cout << "a.reshape(3,2):\n" << a << std::endl;

    vpArray2D<float> a2;
    a2.resize(2,2);
    a2 = {1, 2, 3, 4};
    std::cout << "a2:\n" << a2 << std::endl;

    vpArray2D<double> a3(2, 3, { 1, 2, 3, 4, 5, 6 });
    std::cout << "a3:\n" << a3 << std::endl;

    vpArray2D<int> a4{ {1,2,3},{4,5,6},{7,8,9} };
    std::cout << "a4:\n" << a4 << std::endl;

    vpArray2D<int> a5;
    a5 = { {1,2,3},{4,5,6},{7,8,9} };
    std::cout << "a5:\n" << a5 << std::endl;

    vpArray2D<int> a6{a5};
    std::cout << "a6:\n" << a6 << std::endl;

    vpMatrix m{1, 2, 3};
    std::cout << "m:\n" << m << std::endl;
    m = { -1, -2, -3, -4 };
    std::cout << "m:\n" << m << std::endl;
    m.reshape(2, 2);
    std::cout << "m:\n" << m << std::endl;

    vpMatrix m2(3, 2, { 1, 2, 3, 4, 5, 6 });
    std::cout << "m2:\n" << m2 << std::endl;

    vpMatrix m3{ {1,2,3},{4,5,6},{7,8,9} };
    std::cout << "m3:\n" << m3 << std::endl;

    vpMatrix m4;
    m4 = { {1,2,3},{4,5,6},{7,8,9} };
    std::cout << "m4:\n" << m4 << std::endl;

    vpMatrix m5{m4};
    std::cout << "m5:\n" << m5 << std::endl;

    vpMatrix m6;
    m6 = {m2};
    std::cout << "m6:\n" << m6 << std::endl;

    vpColVector c{ 1, 2, 3 };
    std::cout << "c:\n" << c << std::endl;
    c = { -1, -2, -3 };
    std::cout << "c:\n" << c << std::endl;

    vpRowVector r{ 1, 2, 3, 4, 5, 6 };
    std::cout << "r:\n" << r << std::endl;
    r = { -1, -2, -3 };
    std::cout << "r:\n" << r << std::endl;
  }
#endif

  {
    vpMatrix m1;
    m1 << 1, 2, 3;
    std::cout << "m1:\n" << m1 << std::endl;

    m1 << -1, -2, -3, -4;
    m1.reshape(1, 4);
    std::cout << "m1:\n" << m1 << std::endl;

    vpMatrix m2(2,2);
    m2 << 1, 2, 3, 4, 5, 6, 7, 8, 9;
    std::cout << "m2:\n" << m2 << std::endl;

    m2.resize(3, 3, false);
    std::cout << "m2:\n" << m2 << std::endl;

    m2 << 0.0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11;
    m2.reshape(2, 6);
    std::cout << "m2:\n" << m2 << std::endl;

    vpColVector c;
    c << 1, 2, 3, 4;
    std::cout << "c: " << c << std::endl;
```

### CODE REVIEW

**Comment by fspindle** [comment]
@s-trinh I agree, but in the following snippet extracted from your test, the last line doesn't make sense to me:
```
    vpColVector c;
    c << 1, 2, 3, 4;
    std::cout << "c:\n" << c << std::endl;

    try {
      c.reshape(2, 2);
      std::cout << "after c.reshape(2, 2): " << c.t() << std::endl; <-- here c remains unchanged. This looks good to me
      c = c.reshape(2, 2);
```

It should be rather:
```
    vpColVector c;
    c << 1, 2, 3, 4;
    std::cout << "c:\n" << c << std::endl;

    try {
      c.reshape(2, 2);
      std::cout << "after c.reshape(2, 2): " << c.t() << std::endl;
      vpMatrix m3 = c.reshape(2, 2);
      std::cout << "m3:\n" << m3 << std::endl;
```

**Comment by s-trinh** [comment]
@fspindle 
This was a test to see if somehow it is possible to change the layout of a `vpColVector` or a `vpRowVector`.
An exception is correctly thrown.

With inheritance, it is possible to do it (should not be a very common code).

File: modules/core/include/visp3/core/vpArray2D.h [hunk 1]
```diff
@@ -54,7 +54,7 @@
   \class vpArray2D
   \ingroup group_core_matrices
 
-  \brief Implementation of a generic 2D array used as vase class of matrices
+  \brief Implementation of a generic column-major 2D array used as vase class of matrices
```
Comment (s-trinh) [inline modules/core/include/visp3/core/vpArray2D.h]:
```suggestion
  \brief Implementation of a generic 2D array used as base class for matrices
```
Probably `column-major` means here a matrix represented as a 1D array in memory.
`column-major` looks confusing in my opinion, with Fortran vs C++ column-major vs row-major order.

File: modules/core/include/visp3/core/vpColVector.h [hunk 2]
```diff
@@ -68,6 +68,52 @@ class vpPoseVector;
   operations on these vectors.
 
   The vpColVector class is derived from vpArray2D<double>.
+
+  The code below shows how to create a 3-dim column vector of doubles, set the element values and access them:
```
Comment (s-trinh) [inline modules/core/include/visp3/core/vpColVector.h]:
```suggestion
  The code below shows how to create a 3-element column vector of doubles, set the element values and access them:
```
`dim` is more related to matrix dimension for me.

File: modules/core/include/visp3/core/vpMatrix.h [hunk 3]
```diff
@@ -93,11 +93,71 @@ class vpForceTwistMatrix;
   - If none of these previous 3rd parties is installed, we use by default a
   Lapack built-in version.
 
-  vpMatrix class provides a data structure for the matrices as well
+  vpMatrix class provides a column-major data structure for the matrices as well
```
Comment (s-trinh) [inline modules/core/include/visp3/core/vpMatrix.h]:
```suggestion
  vpMatrix class provides a data structure for the matrices as well
```

File: modules/core/include/visp3/core/vpRowVector.h [hunk 4]
```diff
@@ -68,6 +67,50 @@ class vpColVector;
   these vectors.
 
   The vpRowVector class is derived from vpArray2D<double>.
+
+  The code below shows how to create a 3-dim row vector of doubles, set the element values and access them:
```
Comment (s-trinh) [inline modules/core/include/visp3/core/vpRowVector.h]:
```suggestion
  The code below shows how to create a 3-element row vector of doubles, set the element values and access them:
```

File: modules/core/src/math/transformation/vpTranslationVector.cpp [hunk 5]
```diff
@@ -505,6 +505,96 @@ vpTranslationVector &vpTranslationVector::operator=(double x)
   return *this;
 }
 
+#ifdef VISP_HAVE_CXX11
+/*!
+  Set vector from a list of 3 double angle values in meters.
```
Comment (s-trinh) [inline modules/core/src/math/transformation/vpTranslationVector.cpp]:
```suggestion
  Set vector from a list of 3 double values in meters.
```

**Pull request merged**